File: src/client/java/set/starlev/starredheltix/util/updater/ModUpdater.java
package set.starlev.starredheltix.util.updater;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.sound.PositionedSoundInstance;
import net.minecraft.text.Text;
import set.starlev.starredheltix.sound.ModSounds;

import java.awt.Desktop;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.CompletableFuture;

public class ModUpdater {
    private static final String GITHUB_API_URL = "https://api.github.com/repos/Starlevka/StarredHeltix/releases/latest";
    private static final String CURRENT_VERSION = getCurrentVersion();
    private static String downloadUrl = null;
    private static String latestVersion = null;
    
    public static void checkForUpdates() {
        CompletableFuture.runAsync(() -> {
            try {
                String response = makeHttpRequest(GITHUB_API_URL);
                JsonObject release = JsonParser.parseString(response).getAsJsonObject();
                
                latestVersion = release.get("tag_name").getAsString();
                
                if (isNewerVersion(latestVersion, CURRENT_VERSION)) {
                    // Find .jar asset
                    JsonArray assets = release.getAsJsonArray("assets");
                    for (int i = 0; i < assets.size(); i++) {
                        JsonObject asset = assets.get(i).getAsJsonObject();
                        String name = asset.get("name").getAsString();
                        if (name.endsWith(".jar")) {
                            downloadUrl = asset.get("browser_download_url").getAsString();
                            break;
                        }
                    }
                    
                    MinecraftClient client = MinecraftClient.getInstance();
                    if (client != null && client.player != null) {
                        client.execute(() -> {
                            client.player.sendMessage(Text.literal("§6[StarredHeltix] §eДоступна новая версия: " + latestVersion + "! Хотите установить? /starredheltix update install"), false);
                            playUpdateSound();
                        });
                    }
                } else {
                    MinecraftClient client = MinecraftClient.getInstance();
                    if (client != null && client.player != null) {
                        client.execute(() -> {
                            client.player.sendMessage(Text.literal("§a[StarredHeltix] У вас уже последняя версия: " + CURRENT_VERSION), false);
                        });
                    }
                }
            } catch (Exception e) {
                MinecraftClient client = MinecraftClient.getInstance();
                if (client != null && client.player != null) {
                    client.execute(() -> {
                        client.player.sendMessage(Text.literal("§c[StarredHeltix] Ошибка проверки обновлений: " + e.getMessage()), false);
                    });
                }
            }
        });
    }
    
    public static void downloadUpdate() {
        if (downloadUrl == null) {
            MinecraftClient client = MinecraftClient.getInstance();
            if (client != null && client.player != null) {
                client.player.sendMessage(Text.literal("§c[StarredHeltix] Сначала проверьте обновления!"), false);
            }
            return;
        }
        
        CompletableFuture.runAsync(() -> {
            try {
                MinecraftClient client = MinecraftClient.getInstance();
                if (client != null && client.player != null) {
                    client.execute(() -> {
                        client.player.sendMessage(Text.literal("§e[StarredHeltix] Скачивание обновления..."), false);
                    });
                }
                
                // Download new version
                Path modsDir = Paths.get("mods");
                String fileName = "starredheltix-" + latestVersion + ".jar";
                Path newModPath = modsDir.resolve(fileName);
                
                downloadFile(downloadUrl, newModPath.toString());
                
                // Schedule old version deletion on game exit
                scheduleOldVersionDeletion();
                
                if (client != null && client.player != null) {
                    client.execute(() -> {
                        client.player.sendMessage(Text.literal("§a[StarredHeltix] Обновление скачано! Перезапустите игру для применения."), false);
                    });
                }
                
            } catch (Exception e) {
                MinecraftClient client = MinecraftClient.getInstance();
                if (client != null && client.player != null) {
                    client.execute(() -> {
                        client.player.sendMessage(Text.literal("§c[StarredHeltix] Ошибка скачивания: " + e.getMessage()), false);
                    });
                }
            }
        });
    }
    
    public static void openReleasesPage() {
        try {
            Desktop.getDesktop().browse(URI.create("https://github.com/Starlevka/StarredHeltix/releases"));
        } catch (Exception e) {
            MinecraftClient client = MinecraftClient.getInstance();
            if (client != null && client.player != null) {
                client.player.sendMessage(Text.literal("§c[StarredHeltix] Не удалось открыть браузер"), false);
            }
        }
    }
    
    public static void openProjectPage() {
        try {
            Desktop.getDesktop().browse(URI.create("https://github.com/Starlevka/StarredHeltix"));
        } catch (Exception e) {
            MinecraftClient client = MinecraftClient.getInstance();
            if (client != null && client.player != null) {
                client.player.sendMessage(Text.literal("§c[StarredHeltix] Не удалось открыть браузер"), false);
            }
        }
    }
    
    private static void playUpdateSound() {
        MinecraftClient client = MinecraftClient.getInstance();
        if (client != null) {
            client.getSoundManager().play(PositionedSoundInstance.master(ModSounds.UPDATE_AVAILABLE, 1.0F, 1.0F));
        }
    }
    
    private static String makeHttpRequest(String urlString) throws Exception {
        URL url = new URL(urlString);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("User-Agent", "StarredHeltix-Updater");
        
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        
        return response.toString();
    }
    
    private static void downloadFile(String urlString, String fileName) throws Exception {
        URL url = new URL(urlString);
        try (InputStream in = url.openStream();
             FileOutputStream out = new FileOutputStream(fileName)) {
            byte[] buffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = in.read(buffer)) != -1) {
                out.write(buffer, 0, bytesRead);
            }
        }
    }
    
    private static boolean isNewerVersion(String latest, String current) {
        String[] latestParts = latest.replaceAll("[^0-9.]", "").split("\\.");
        String[] currentParts = current.replaceAll("[^0-9.]", "").split("\\.");
        
        int maxLength = Math.max(latestParts.length, currentParts.length);
        for (int i = 0; i < maxLength; i++) {
            int latestPart = i < latestParts.length ? Integer.parseInt(latestParts[i]) : 0;
            int currentPart = i < currentParts.length ? Integer.parseInt(currentParts[i]) : 0;
            
            if (latestPart > currentPart) return true;
            if (latestPart < currentPart) return false;
        }
        return false;
    }
    
    private static String getCurrentVersion() {
        try {
            InputStream stream = ModUpdater.class.getResourceAsStream("/version.txt");
            if (stream != null) {
                BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
                String version = reader.readLine();
                reader.close();
                return version != null ? version.trim() : "0.0.5";
            }
        } catch (Exception e) {
            // Fallback
        }
        return "0.0.5";
    }
    
    private static void scheduleOldVersionDeletion() {
        // Простое и безопасное решение - просто показываем сообщение
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            MinecraftClient client = MinecraftClient.getInstance();
            System.out.println("[StarredHeltix] Обновление завершено! Не забудьте удалить старую версию мода из папки mods/");
        }));
    }
}

File: src/main/kotlin/set/starlev/utils/ModUpdater.kt
package set.starlev.utils

import com.google.gson.JsonParser
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.client.Minecraft
import net.minecraft.network.chat.Component
import set.starlev.ModSounds
import java.net.HttpURLConnection
import java.net.URL
import java.nio.file.Files
import java.nio.file.Paths
import java.util.concurrent.CompletableFuture

object ModUpdater {
    private val mc = Minecraft.getInstance()
    private const val GITHUB_API = "https://api.github.com/repos/Starlevka/StarredHeltix/releases/latest"
    private var downloadUrl: String? = null
    private var latestVersion: String? = null
    
    private val currentVersion: String
        get() = FabricLoader.getInstance().getModContainer("starredheltix")
            .map { it.metadata.version.friendlyString }
            .orElse("0.0.8")

    fun checkUpdate(quiet: Boolean = false) {
        CompletableFuture.runAsync {
            try {
                val response = makeRequest(GITHUB_API)
                val json = JsonParser.parseString(response).asJsonObject
                
                latestVersion = json.get("tag_name").asString.removePrefix("v")
                
                if (isNewer(latestVersion!!, currentVersion)) {
                    json.getAsJsonArray("assets").forEach { asset ->
                        val obj = asset.asJsonObject
                        if (obj.get("name").asString.endsWith(".jar")) {
                            downloadUrl = obj.get("browser_download_url").asString
                        }
                    }
                    
                    mc.execute {
                        if (!quiet) {
                            mc.player?.displayClientMessage(
                                Component.literal("§6[StarredHeltix] §eНовая версия: $latestVersion! §7/starredheltix update install"),
                                false
                            )
                            mc.player?.playSound(ModSounds.UPDATE_AVAILABLE, 1.0f, 1.0f)
                        } else {
                            // Если тихо (при входе), показываем только если есть обнова
                            // Добавляем небольшую задержку, чтобы сообщение не потерялось в спаме при входе
                            CompletableFuture.runAsync {
                                try { Thread.sleep(3000) } catch (e: Exception) {}
                                mc.execute {
                                    mc.player?.displayClientMessage(
                                        Component.literal("§6§l[StarredHeltix] §e§lДоступно обновление: §a§l$latestVersion! §7(Текущая: $currentVersion)"),
                                        false
                                    )
                                    mc.player?.displayClientMessage(
                                        Component.literal("§7Используйте §f/sh update install §7для автоматической установки."),
                                        false
                                    )
                                    mc.player?.playSound(ModSounds.UPDATE_AVAILABLE, 1.0f, 1.0f)
                                }
                            }
                        }
                    }
                } else if (!quiet) {
                    mc.execute {
                        mc.player?.displayClientMessage(
                            Component.literal("§a[StarredHeltix] Последняя версия: $currentVersion"),
                            false
                        )
                    }
                }
            } catch (e: Exception) {
                if (!quiet) {
                    mc.execute {
                        mc.player?.displayClientMessage(
                            Component.literal("§c[StarredHeltix] Ошибка проверки: ${e.message}"),
                            false
                        )
                    }
                }
            }
        }
    }

    fun installUpdate() {
        if (downloadUrl == null) {
            mc.player?.displayClientMessage(
                Component.literal("§c[StarredHeltix] Сначала проверьте обновления!"),
                false
            )
            return
        }
        
        CompletableFuture.runAsync {
            try {
                mc.execute {
                    mc.player?.displayClientMessage(
                        Component.literal("§e[StarredHeltix] Скачивание..."),
                        false
                    )
                }
                
                val modsDir = Paths.get("mods")
                val fileName = "starredheltix-$latestVersion.jar"
                val newMod = modsDir.resolve(fileName)
                
                java.net.URI(downloadUrl!!).toURL().openStream().use { input ->
                    Files.copy(input, newMod)
                }
                
                mc.execute {
                    mc.player?.displayClientMessage(
                        Component.literal("§a[StarredHeltix] Мод установлен! §cУдалите старую версию из папки mods и перезапустите игру."),
                        false
                    )
                }
            } catch (e: Exception) {
                mc.execute {
                    mc.player?.displayClientMessage(
                        Component.literal("§c[StarredHeltix] Ошибка: ${e.message}"),
                        false
                    )
                }
            }
        }
    }

    private fun makeRequest(url: String): String {
        val conn = java.net.URI(url).toURL().openConnection() as HttpURLConnection
        conn.requestMethod = "GET"
        conn.setRequestProperty("User-Agent", "StarredHeltix-Updater")
        return conn.inputStream.bufferedReader().use { it.readText() }
    }

    private fun isNewer(latest: String, current: String): Boolean {
        val latestParts = latest.replace(Regex("[^0-9.]"), "").split(".")
        val currentParts = current.replace(Regex("[^0-9.]"), "").split(".")
        
        for (i in 0 until maxOf(latestParts.size, currentParts.size)) {
            val l = latestParts.getOrNull(i)?.toIntOrNull() ?: 0
            val c = currentParts.getOrNull(i)?.toIntOrNull() ?: 0
            if (l > c) return true
            if (l < c) return false
        }
        return false
    }
}
